compiler/src/dotty/tools/dotc/util/Signatures.scala

```diff
@@ -41,109 +43,238 @@ object Signatures {
+   * @param fun Unapply function tree
+   *
+   * @return A triple containing the index of the parameter being edited, the index of the function
+   *         being called, the list of overloads of this function).
+   */
   private def unapplyCallInfo(
     span: Span,
     fun: tpd.Tree,
     patterns: List[tpd.Tree]
-  )(using Context): (Int, Int, List[SingleDenotation]) =
-    val patternPosition = patterns.indexWhere(_.span.contains(span))
-    val activeParameter = extractParamTypess(fun.tpe.finalResultType.widen).headOption.map { params =>
-      (patternPosition, patterns.length) match
-        case (-1, 0) => 0 // there are no patterns yet so it must be first one
-        case (-1, pos) => -1 // there are patterns, we must be outside range so we set no active parameter
-        case _ => (params.size - 1) min patternPosition max 0 // handle unapplySeq to always highlight Seq[A] on elements
-    }.getOrElse(-1)
+  )(using Context): (Int, Int, List[Signature]) =
+    val resultType = unapplyMethodResult(fun)
+    val isUnapplySeq = fun.denot.name == core.Names.termName("unapplySeq")
+    val denot = fun.denot
+
+    def extractParamTypess(resultType: Type): List[List[Type]] =
+      resultType match
+        // Reference to a type which is not a type class
+        case ref: TypeRef if !ref.symbol.isPrimitiveValueClass => mapOptionLessUnapply(ref, patterns.size, isUnapplySeq)
+        // Option or Some applied type. There is special syntax for multiple returned arguments: Option[TupleN]
+        // We are not intrested in it, instead we extract proper type parameters from the Option type parameter.
+        case AppliedType(TypeRef(_, cls), (appliedType @ AppliedType(tycon, args)) :: Nil)
+            if (cls == ctx.definitions.OptionClass || cls == ctx.definitions.SomeClass) =>
+          tycon match
+            case typeRef: TypeRef if ctx.definitions.isTupleClass(typeRef.symbol) => List(args)
+            case _ => List(List(appliedType))
+        // Applied type extractor. We must extract from applied type to retain type parameters
+        case appliedType: AppliedType => mapOptionLessUnapply(appliedType, patterns.size, isUnapplySeq)
+        // This is necessary to extract proper result type as unapply can return other methods eg. apply
+        case MethodTpe(_, _, resultType) => extractParamTypess(resultType.widenDealias)
+        case _ => Nil
+
+    def extractParamNamess(resultType: Type): List[List[Name]] =
+      if resultType.typeSymbol.flags.is(Flags.CaseClass) && denot.symbol.flags.is(Flags.Synthetic) then
+        resultType.typeSymbol.primaryConstructor.paramInfo.paramNamess
+      else
+        Nil
+
+    val paramTypes = extractParamTypess(resultType).flatten.map(stripAllAnnots)
+    val paramNames = extractParamNamess(resultType).flatten
+
+    val activeParameter = unapplyParameterIndex(patterns, span, paramTypes.length)
+    val unapplySignature = toUnapplySignature(paramNames, paramTypes).toList
```

Comment:
This line required signature change, we had to distinguish apply from unapply to fix 
```scala object And {
  def unapply[A](a: A): Some[(A, A)] = Some((a, a))
}
object a {
  And.unapply(@@)
}
```